### DISCUSSION TITLE
Proposing state management capabilities to NCronJob

### DISCUSSION THREAD

**Discussion by falvarez1** [NCronJob-Dev/NCronJob#25] [Ideas]
I'd like to discuss the possibility of enhancing `NCronJob` with state management for jobs, similar to other job orchestration frameworks. This feature would track states like `Scheduled`, `Running`, `Retrying`, `Completed`, `Failed`, and `Cancelled`. Here are the potential benefits and features:

- **Enhanced Observability**: Track and visualize job states in real-time, providing better insights and monitoring capabilities.
- **Actionable States**: Implement state change hooks that allow actions to be triggered based on state transitions. This can be used for notifications, cleanup tasks or conditional execution of other jobs (job chaining).
- **Job History and Persistence**: With state tracking, we can store historical data on job executions, which can be useful for auditing, debugging, and optimizing job schedules.

This enhancement would not only improve the robustness and usability of `NCronJob` but also bring it closer to feature parity with more advanced workflow orchestration tools.

I see state management as the core just the beginning of what we can do. Starting with basic state management, we have the opportunity to lay a solid foundation that future enhancements can build upon. We can kick things off with the essentials, and then, based on our experiences and feedback, we can expand and refine the features over time.

Would love to hear your thoughts on this idea.

**Comment by linkdotnet** [comment]
Thanks for the rich ideas!

* **Enhanced Observability**: I am totally on board! Currently, it is more of a black box that isn't created in somewhat enterprise-grade applications.
* **Actionable States**: In general, I like the idea. My gut tells me that it is a slippery slope. On the one hand, I like to keep the scheduler rather simple than complex (otherwise, it might end up as yet another Quartz/Hangfire clone). On the other hand, I do understand the need for that. To make it right there has to be a lot of flexibility in the system as well as a nice API that keeps it nice and easy for you
* **Job History and Persistence**: Also one of the things I ruled out, with the same explanation as above. I wanted to have a very low barrier to use the library with almost 0 config.

But all in all these are very good features. My current way of thinking is to make the **NCronJob** library Plug&Play ready and offer hooks for extensions packages.

For example, **persistence**: We would probably need multiple providers (like Postgres, SQL Server, Mongo, Cosmos, ...), so the system has to be flexible either way.

**Reply by falvarez1** [reply]
For actionable states, we certainly could hash out the design details here.
This wouldn't be another Quartznet/Hangfire. TBH every time I've tried using them I've always found something missing and would end up writing something custom or finding other solutions.

I hear you on very low barrier and easy solution and I'm on the same page. Persistence doesn't necessarily have to use external databases like you mentioned. In order to start simple, or even just keep things simple, we could use a file based write-ahead logging approach. Zero config to the end user. Maybe just a switch that says "persist to disk".
Microsoft has a project called [FASTER](https://github.com/microsoft/FASTER) that may be an option. Either way I think start simple, no need to set up docker-compose scripts with services that call to databases etc. just to set up the infrastructure.

**Reply by linkdotnet** [reply]
https://github.com/microsoft/FASTER sounds nice! I am very much in favor of just trying it out and in the worst scrap it again.
Alone the name of the library should be a reason to use it :D

Kidding aside - I do agree that is a good start. I had some previous exposure to job-servers in the .net landscape and being overly flexible comes with a high price. That is why I wanted to keep it as simple as possible. So yeah! You convinced me to try out this direciton.

**Comment by falvarez1** [comment]
@linkdotnet 
I started a branch at [experiment/state-events-system](https://github.com/linkdotnet/NCronJob/tree/experiment/state-events-system) with some preliminary state management. It's still a WIP but I'd like your feedback if you get a chance.

**Reply by linkdotnet** [reply]
Interesting - I do like the Pub/Sub Part and do see the potential.

I know it is a PoC/Experiment at the moment, but maybe a small question:
In the `JobExecutor` you are setting state - wouldn't it be easier just to send messages from various classes (like Scheduler/JobExectuor/RetryHandler/...) and have a central entity handling that message?

**Reply by falvarez1** [reply]
I tried that and had some issues with ordering of states. Nothing that can't be solved. I'm not super happy about how state is defined at the moment. I feel like state should be inherently part of a state machine, right now it's just a side effect of a process outside of any state machine. There's potential to get out of sync with the correct representation of state since behavior and the structure is decoupled.
It's still very experimental so I'll try a few different approaches.

**Reply by linkdotnet** [reply]
Good idea. The central approach could be like this:
```mermaid
classDiagram
    JobScheduler <-->  Mediator
    JobExecutor <-->  Mediator
    Mediator <--> JobObserver
end
```

The `Mediator` is very central that mediates all the events - with that we can control the order of events without a problem.
That would go hand in hand with the Pub/Sub approach you described earlier.

Stuff like retrying and so on would also result in "messages" handled by the Mediator in the middle.
I don't expect that this object would handle thousands of messages a second, so we don't create a bottleneck here.

**Reply by falvarez1** [reply]
If I understand correctly, you're suggesting the Mediator handle both event publishing and command issuance for actions such as trigger job, cancel job, start/stop scheduler etc. This could centralize job management and streamline notifications. Introducing a unified Mediator also opens up the possibility to consolidate job commands in a service interface, which can be exposed via API endpoints to enhance programmability and external interaction with the system.

I've previously thought about merging these functions into a single Mediator, but it could significantly alter the current architecture. Currently, the `JobStateManager` functions as a rudimentary Mediator, with the `MessageHub` handling event distribution. While this setup avoids bottlenecks and manages events effectively, it may restrict control over event sequencing and command distribution.

I'm open to experiment with a unified Mediator to assess its impact on system performance and maintainability. I plan to integrate updates gradually and would appreciate your insights as we progress.

**Reply by linkdotnet** [reply]
> If I understand correctly, you're suggesting the Mediator handle both event publishing and command issuance for actions such as trigger job, cancel job, start/stop scheduler etc

Yes - but I am currently going through rather or not that covers all the cases and is logical to follow.

> Introducing a unified Mediator also opens up the possibility to consolidate job commands in a service interface, which can be exposed via API endpoints to enhance programmability and external interaction with the system.

Exactly - I tried to some rough sketches in code where there was an almighty Mediator, which creates a bit of a mess inside that class. But it leaves us the option of just "adding" more observers as we wish.

PS: The JobObserver should only be notified and don't push stuff towards the Mediator - that's wrong in the sketch above.

**Reply by linkdotnet** [reply]
I am thinking of something like this:
```mermaid
sequenceDiagram
    participant A as CronScheduler (BackgroundService)
    participant B as Mediator
    participant C as JobExecutor
    participant D as ObserverA
    participant E as RetryHandler
    A-->>B: Notify about new jobs scheduled
    B-->>D: Notify
    A-->>B: Notify that one job has to be executed now<br/>and reschedule
    B-->>D: Notify
    B-->>C: Pass job for exeuction
    B-->>D: Notify
    alt Job successful
      C-->>B: Notify Job exeuction successful
      B-->>D: Notify
    end
    alt Job faulted
      C-->>B: Notify Job exeuction faulted
      B-->>D: Notify
      C-->>E: Ask for retry
      E-->>C: Retry Nr. 1
      B-->>D: Notify
      B-->>C: Pass job for exeuction
      B-->>D: Notify
    end
```

Alternatively we distribute some of the logic (I will leave out external observers as they introduce a bit too much noise):
```mermaid
sequenceDiagram
    participant A as CronScheduler
    participant B as Mediator
    participant C as JobExecutor
    participant D as CronRegistry

    D->>B: Informs about CRON jobs (or instant jobs)
    alt Schedule CRON Job
        B->>B: Calculate next run time based on CRON expression (internal)
        B-->A: Update Job Queue (with RegistryEntry containing job details and next run time)
    else Schedule Instant Job
        B->>A: Add job to queue immediately
    end
    loop
        A->>B: Get next job from queue
        alt Success
            B->>A: Job details (RegistryEntry)
            A->>C: Execute job (using JobExecutor)
            C->>A: Job completed (implicit)
            A->>B: Get next job from queue (loop back)
        else Job queue empty
            A->>B: Check for cancellation
        end
    end
```

Basically everytime something goes into the `Mediator` we would call the observers with their appropriate notification.

**Reply by linkdotnet** [reply]
> I'm open to experiment with a unified Mediator to assess its impact on system performance and maintainability. I plan to integrate updates gradually and would appreciate your insights as we progress.

Cool! My way of thinking is that there aren't many<sup>tm</sup> jobs in parallel scheduled. There could be many in execution, but that is "fine".

It might become an issue if we allow the Job itself to communicate with the outside world via an interface to express certain state-changes. For example:
```csharp
public class MyJob : IJob
{
  public void RunAsync(JobExecutionContext context, CancellationToken token)
  {
    await DoStep1Async();
    context.Notifier.SetProgress(0.3); // Range of 0..1
    await DoStep2Async();
    context.Notifier.SetProgress(0.6);
    await DoStep3Async();
  }
}
```

Basically we are offering a service that allows "external" communication - the use case would be dashboard related.
Imagine you want to see your jobs in "real life" - With that, we can easily have progress bars and whatnot.

**Reply by linkdotnet** [reply]
Sorry for spamming the discussion here - there are many ideas, and I guess for now, we have to settle on one thing.
My preferred way is a single unified Mediator. If we see no fit for this anymore, we have to move it into different components.

**Reply by falvarez1** [reply]
Yes the idea was to broadcast or stream the state changes and progress to a dashboard. I don't think this would be too heavy so long as we buffer the events. Instead of sending one at a time you aggregate all events over a period, let's say one second, then flush everything in the buffer. I can think of a few other strategies to deal with this at various levels of complexity but keeping it simple is best.

**Reply by linkdotnet** [reply]
Another concept that somewhat fits "natural" here is RxJS / RxNET with a store approach.
Given our requirements for buffering and whatnot, it could be worth a shot.

The only issue is that folks normally don't like it :D

**Reply by falvarez1** [reply]
Yes it does fit well and is a great package. I've used Reactive extensions many years ago, on different platforms, however I tend to shy away from heavy packages unless absolutely necessary. I think basic buffering can be done with standard data structures.

**Reply by falvarez1** [reply]
I've used this approach in other projects successfully (simplified here to represent the concept):

```csharp
public class BufferedNotifier<T> : IDisposable
{
    private readonly IMessageService<T> _messageService;
    private readonly List<T> _messageBuffer = new List<T>();
    private readonly Timer _timer;
    private readonly SemaphoreSlim _asyncLock = new SemaphoreSlim(1, 1);
    private readonly ILogger<BufferedNotifier<T>> _logger;

    public BufferedNotifier(IMessageService<T> messageService, ILogger<BufferedNotifier<T>> logger)
    {
        _messageService = messageService;
        _logger = logger;
        _timer = new Timer(async _ => await FlushMessagesAsync(), null, Timeout.Infinite, Timeout.Infinite);
    }

    public async void Notify(T message)
    {
        await HandleMessageAsync(message);
    }

    private async Task HandleMessageAsync(T message)
    {
        _logger.LogTrace("Handling message");

        if (!_messageService.IsConnected)
        {
            _logger.LogWarning("The message service connection is not open.");
            return;
        }

        await _asyncLock.WaitAsync().ConfigureAwait(false);
        try
        {
            _messageBuffer.Add(message);

            if (_messageBuffer.Count == 1)
            {
                _timer.Change(TimeSpan.FromSeconds(1), Timeout.InfiniteTimeSpan);
            }
        }
        finally
        {
            _asyncLock.Release();
        }
    }

    private async Task FlushMessagesAsync()
    {
        await _asyncLock.WaitAsync();
        List<T> messagesToSend;

        try
        {
            if (_messageBuffer.Count == 0)
                return;

            messagesToSend = [.._messageBuffer];
            _messageBuffer.Clear();
        }
        finally
        {
            _asyncLock.Release();
        }

        if (_messageService.IsConnected)
        {
            try
            {
                await _messageService.SendBatchMessageAsync("ReceiveMessagesBatch", messagesToSend);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "An error occurred while sending batched messages");
            }
        }
    }

    public void Dispose()
    {
        _timer?.Dispose();
        _asyncLock?.Dispose();
    }
}
```

**Reply by linkdotnet** [reply]
> Yes it does fit well and is a great package. I've used Reactive extensions many years ago, on different platforms, however I tend to shy away from heavy packages unless absolutely necessary. I think basic buffering can be done with standard data structures.

Fair point. Yes - it doesn't necessarily call for such heavy calibers here. Once you worked with such libraries the flow of data is somewhat nice to follow.

But for now - let's start with a unified mediator.

**Comment by falvarez1** [comment]
@linkdotnet I created this branch to start working on designing the job commands capabilities to support the actions that would be performed on the Dashboard. So far supports: 

- Cancel in-progress JobRun
- Pause JobRun
- Resume JobRun
- Pause Job Queue
- Resume Job Queue
- Cancel all in-progress Job Runs for a given Job type

[Job Commands Management WIP branch
](https://github.com/NCronJob-Dev/NCronJob/tree/job-management)

Currently focusing on making sure the basic commands work well.
Looking for feedback

**Reply by linkdotnet** [reply]
It's a really nice start - I was wondering what "Pause JobRun" should do.

By the way that would be also a very nice enabler for the whole persistence story (if we do that).

**Reply by falvarez1** [reply]
I guess technically it's more of a "pause Job" or "pause job triggers", where a trigger can be a scheduled execution, manual trigger (instant job), and anything else that may start execution. The act of pausing a Job means you don't want any job of that Job Type to be executed. There's 3 options I see here for what it means:

- Option 1: Removing Paused Jobs from the Queue - When a job is paused, any jobs of that type currently in the queue are removed. This ensures that no paused jobs will be executed, but it also means that those jobs need to be re-queued when the job is resumed. State is persisted in the `JobRegistry`
- Option 2: Mark Jobs as Paused in the Queue - When a job is paused, any jobs of that type currently in the queue are marked as paused. When it comes time to execute, these jobs are skipped. State is persisted in the `JobRunManager` or the `JobRun`
- Option 3: Pausing the Job Queue - The entire queue stops processing jobs. Jobs remain in the queue, but they are not dequeued or executed until the queue is resumed. This may create a tighter coupling between a Job and a Queue. The state would be defined and persisted in the `JobQueue` so that the `JobWorker` can use that as the reference.

There's a few design decisions that need to be made. Depending on the option we may need a higher level _JobType_ which itself contains the list of instances (JobDefinitions). Or similarly to how a `JobRun` references a `JobDefinition`, we can have a `JobDefinition` reference a _JobType_. This _JobType_ would have the metadata that doesn't change like policy metadata, JobName, JobFullName, and could include a _job type state_ with the values (Active, Disabled, Paused) and its own Id (for persistence). We may need this for persistence either way.

**Reply by linkdotnet** [reply]
Option 1 and 3 sounds exactly like the "Disable Job" function we introduced with the runtime feature some days ago.
And Option 2 is like a variation of that. So, if we introduce something along the lines of Option 1, we might reconsider (for v3) what our public surface will be.

Of course any of those options above are somewhat better as we currently "trick" the system with an invalid CRON expression.
The runtime registry can use the manager under the hood to fulfill the requirements.

**Reply by falvarez1** [reply]
Great point! So option 2 seems to be the best representation of a paused job. And seems to align with other libraries like Quartz. I'll draft up some assumptions we can review.

**Reply by falvarez1** [reply]
@linkdotnet I’m jumping back into this! I see there have been some changes since I last worked on the code, so I’ll need to review them. I’d love to get the dashboard finalized and fully functional. Time to dust off my dashboard branch and get cracking! 🚀

**Reply by linkdotnet** [reply]
Hey, welcome back. Unfortunately, I noticed your message was a bit too late. 

The majority of changes are in test-setup. @nulltoken added some observability, which made the Channel approach we used earlier obsolete. We discovered some issues with `IDisposable` and made smaller changes to the API surface as well. The majority of work was done by @nulltoken - I don't want to take much credit here :D.

At its core, the behavior is somewhat the same (hopefully :D)

**Reply by nulltoken** [reply]
@falvarez1 Some additional context:

Observability was added to scratch a personal itch (#147).

It eventually allowed to tweak the way we write tests (some remaining work still in flight in #209 & #199).

Observability feature isn't done yet and `ExecutionProgress` will need to be enhanced to make #36 possible.

@linkdotnet, for instance, was willing to potentially surface the `IJobExecutionContext` through it. However, in my mind, the `ExecutionProgress` should be an immutable snapshot (as a consumer of the IObserver interface, I wouldn't want a process updating previous states behind my back ;-)).

As I needn't it yet, I deferred that work to finish updating the test suite. However, we will eventually need new props exposed (but maybe a little brain work should be done before in order to define what _identifies_ a job).

While working on Observability, the concept of `Orchestration` emerged. It's more or less an abstract wrapper around the execution of a job and of all the dependents jobs it's linked to.

```
An orchestration is uniquely identifed by an identifier. All jobs belonging to an orchestration share
this same `CorrelationId`.

From a timeline perspective, an orchestration starts before the root job that initiated it
and completes when **all** of its leaf jobs have reached a final state.
```

As the discussion is about state management, an additional random thought: Currently we're rely on an impossible date to disable jobs (Feb, 31st). Although that works, that looks like a bit of a hack. I wonder if we shouldn't switch to a simple more straightforward internal boolean. Thoughts?

**Reply by linkdotnet** [reply]
Very good summary and overall points @nulltoken 

We had some major discussions around Jobs and state here: #137 

>  However, in my mind, the ExecutionProgress should be an immutable snapshot

Yeah. There are many places where this might be useful but it would increase overall complexity to some degree (for example, Passed-in arguments to a job are completely mutable, which might be surprising).

> I wonder if we shouldn't switch to a simple more straightforward internal boolean

Indeed. But that goes into the same discussion as #137: How do we structure the entities we have? What does what and so on. Currently, the easiest would be that boolean on the JobDefinition, but that would be the quick-fix (as 31st of Feb. is).

PS: @falvarez1 and I started to do a discussion on the Matrix Channel. We could "resurrect" the channel.

**Reply by nulltoken** [reply]
> > However, in my mind, the ExecutionProgress should be an immutable snapshot
> 
> Yeah. There are many places where this might be useful but it would increase overall complexity to some degree (for example, Passed-in arguments to a job are completely mutable, which might be surprising).

Hence, one of the reasons why I deferred the task until now ;-)
That looks like a case of "let's not do it yet. At least, until there's a real use case demanding it"

> We had some major discussions around Jobs and state here: #137

I intend to go back to #137 once I'm done with #209 and #199
